plugin/evm/vm.go

```diff
@@ -1271,6 +1279,50 @@ func (vm *VM) initBlockBuilding() error {
 	return nil
 }
 
+func (vm *VM) SubscribeToEvents(ctx context.Context) commonEng.Message {
+	// Block building is not initialized yet, so we haven't finished syncing or bootstrapping.
+	if vm.builder == nil {
+		select {
+		case <-ctx.Done():
+			return 0
+		case ss := <-vm.stateSyncDone:
+			return ss
+		case <-vm.shutdownChan:
+			return 0
+		case <-vm.finishedSyncing:
+			break
+		}
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	pending := make(chan commonEng.Message, 1)
+
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		pending <- vm.builder.waitForTxEnqueue(ctx)
+	}()
+
+	defer wg.Wait()
+
+	select {
+	case ss := <-vm.stateSyncDone:
+		return ss
+	case pendingTx := <-pending:
+		return pendingTx
+	case <-ctx.Done():
+		vm.builder.pendingSignal.Broadcast()
+		return commonEng.Message(0)
+	case <-vm.shutdownChan:
+		vm.builder.pendingSignal.Broadcast()
+		return commonEng.Message(0)
```

Comment:
I don't think this correctly guarantees that `waitForTxEnqueue` will re-check the select case.

Specifically, it is possible for `waitForTxEnqueue` to have it the `default` case of the `select`, but that goroutine has not yet called `pendingSignal.Wait()`. If we call `pendingSignal.Broadcast()` _prior_ to `pendingSignal.Wait()`, the goroutine will not be awoken (because it isn't sleeping). It will then call `pendingSignal.Wait()` and wait forever.

This can happen because `pendingSignal.Broadcast()` is not called with the `b.buildBlockLock` held.

Reply:
Yep, nice catch! Fixed.